### PR TITLE
Make the default list of plugins bundled with logstash a json file

### PR DESCRIPTION
**Pull request by purbon**
So it can be consumed either by LS or external entities that might be interested to know the list of default plugins (for example the new plugins CI status page)

@ph your thoughts here? this basically implements your proposal to have the logstash-ci status page split re plugins.

### CODE REVIEW

File: rakelib/default_plugins.rb [hunk 1]
```diff
@@ -1,114 +1,10 @@
...
-      logstash-output-s3
-      logstash-output-sns
-      logstash-output-sqs
-      logstash-output-statsd
-      logstash-output-stdout
-      logstash-output-tcp
-      logstash-output-udp
-      logstash-output-kafka
-    )
+    DEFAULT_PLUGINS = JSON.parse(File.read("rakelib/default_plugins.json"))["default_plugins"]
```
Comment (ph) [inline rakelib/default_plugins.rb]: Any reason why don't use our json gems?

Comment (purbon) [inline rakelib/default_plugins.rb]: I went to this because I was not sure we load actually logstash-core when running rake.

**Comment by ph** [comment]
@purbon Lets take the time to drop the content of all the constants from this file in the JSON file. (CORE_SPECS_PLUGINS, TEST_JAR_DEPENDENCIES_PLUGINS, TEST_VENDOR_PLUGINS, ALL_PLUGINS_SKIP_LIST)

I think its better that we don't have to edit multiples places when we are dealing with the plugins, WDTY @suyograo  and @purbon  ?

File: rakelib/default_plugins.rb (hunk 1 above)
Comment (ph) [inline rakelib/default_plugins.rb]: Good point, can you test if its the case?

**Comment by purbon** [comment]
I understand the idea of having this in one file, however we do think this separation is good for now. We add the json file in order to enable others to read it (not just logstash), so:

* Other components for now are not interested to read about CORE_SPECS for example.
* Nearly all this are going to be gone as soon as we improve our testing and packages procedures, so for me they are living zombies for now.
* We do need at the end of the day to read this file and make it accessible from ruby.

and must say I understand your idea of having all in a single place, this is are the reasons why I just keep the default ones there.

File: rakelib/default_plugins.rb (hunk 1 above)
Comment (purbon) [inline rakelib/default_plugins.rb]: works, pushing the change.

**Comment by colinsurprenant** [comment]
Any reason not to have just a (more readable) plain json array, sorted by type and plugin name? in the end this is human-edited, for example:

```json
[
"logstash-codec-collectd",
"logstash-codec-dots",
"logstash-codec-edn",
"logstash-codec-edn_lines",

"logstash-input-heartbeat",

"logstash-output-xmpp",
"logstash-output-zeromq",
]
```

**Comment by colinsurprenant** [comment]
also agree with @ph, if we are to extract one list of plugins, lets be consistent and extract all plugin lists.

**Comment by guyboertje** [comment]
@purbon, @ph, @colinsurprenant:
I can't help feeling this should be a JSON object with the key being the plugin and the value another object e.g. `{"default" : true,"all_plugins_skip_list" : true}`
Plus abstracting the reading and querying of the file into a ruby class that rake tasks or others can use.
Or have several JSON array based files that the ruby class reads - maybe glob for `*_plugins_list.json`.
e.g. `default_plugins_list.json`, `core_specs_plugins_list.json` etc.

Basically isolate the format (and its evolution) of the file(s) behind an API.
If the need for CORE_SPECS_PLUGINS goes away then the API returns `[]` if queried for.

**Comment by purbon** [comment]
@guyboertje this will make the format a big more interesting I see, but end of the day I tried to make it really simple and reduce process in consumers of this file.

**Comment by suyograo** [comment]
+1 on making this a JSON object that @guyboertje mentioned. I'd like to use it to add more metadata, for example, to add a warning blob for unsupported plugins in docs.

**Comment by ph** [comment]
> warning blob for unsupported plugins in docs.
Oh this is a nice idea.

**Comment by purbon** [comment]
Ok, will make it like this.

/purbon

**Comment by purbon** [comment]
Changes pushed to include metadata in the file, opinions welcome.

File: rakelib/default_plugins.rb [hunk 2]
```diff
@@ -168,5 +17,24 @@ def self.is_released?(plugin)
       require 'gems'
       Gems.info(plugin) != "This rubygem could not be found."
     end
+
+    def self.fetch_plugins_for(clazz)
+      LogStash::Json.load(::File.read("rakelib/plugins-metadata.json")).select do |_, metadata|
+        metadata[clazz]
+      end.keys
+    end
```
Comment (ph) [inline rakelib/default_plugins.rb]: Its not really a clazz, more like a type or field?

Comment (purbon) [inline rakelib/default_plugins.rb]: works for me, will update.

- purbon

**Comment by ph** [comment]
@purbon I think @guyboertje meant to encapsulate the logic of dealing with the `plugins-metadata.json` into his own class, to allow us to enrich it with other methods to query information for specific plugins, this would be perfectly handy when we do docs generation.

@guyboertje Would you mind clarifying your idea?

**Comment by purbon** [comment]
sorry, did not get it like this, but the idea is ok for me, even this class
will be mostly a wrapper for what the rakelib file is doing.

**Comment by purbon** [comment]
@ph @guyboertje most of this logic will be just the small function I shared, but will add more if you with in a class, however for now I don't see much of benefit .... unless we move this class to core, what it might make a bit more sense to me. What do you think?

File: rakelib/plugins-metadata.json [hunk 3]
```diff
@@ -0,0 +1,905 @@
+{
+  "logstash-codec-cloudtrail": {
+    "default-plugins": false,
```
Comment (colinsurprenant) [inline rakelib/plugins-metadata.json]: is there particular reason for choosing to list by plugins instead of by types? for example:
```
{
  "default-plugins" : [
    "logstash-codec-dots",
    "logstash-codec-collectd",
    ...
  ],
  "core-specs": [
    ...
  ],
  ...
}
```

**Comment by guyboertje** [comment]
I think it should be a class in core or devutils maybe.
The class is responsible for finding and loading the json file(s).
It should act like a mini database not quite SQL queries but a simple query mechanism.
"All entries having property X"
"All entries having property X and Y"
"All entries having property X or Y"
"All entries not having property X"

Maybe map/reduce - map "All with X" reduce "emit properties M and N"

WDYT?

**Comment by colinsurprenant** [comment]
@guyboertje are we seeing any other properties than being part of a group? Sorry to say be it feels we are over-engineering this very simple use-case which is to externalize the plugins groupings, in json, so that these can be consumed by potential external processes like CI. IMO internally in LS, we could just have a method which returns a hash with group names as keys and list of plugins as values, no?

File: rakelib/plugins-metadata.json (hunk 3 above)
Comment (guyboertje) [inline rakelib/plugins-metadata.json]: @colinsurprenant @ph @purbon - it very much depends on whether the format is designed for humans first or the encapsulating class first.

For example (foolish but bear with me), we could simply have `"logstash-codec-dots" : {"id" : 101, (more properties)}` and some indexes ` "default-plugins" : [101, 104, 110]`
This would be friendly to the class but hard for humans.

**Comment by guyboertje** [comment]
@colinsurprenant - I think it can be what ever we want - from one simple class with a simple file format to ES hosted docs with a front-end.

I had not thought of other systems reading the json files. I only imagined that it would be used in our code or maybe in the terminal via rake.

I guess CI tools could use rake.

**Comment by guyboertje** [comment]
@colinsurprenant - Other properties: "unsupported", "community maintained".

**Comment by suyograo** [comment]
> Other properties: "unsupported", "community maintained".

+1.

**Comment by colinsurprenant** [comment]
> Other properties: "unsupported", "community maintained".

And for what use-case exactly?

I am not sure why we need to go beyond simple grouping here. Plugins already have metadata/properties embedded in their gemspecs. Not sure why we need to duplicate that here for as-of-yet non-existing use cases?

I'm not following.

**Comment by purbon** [comment]
In that case I agree, in part, with @colinsurprenant, the first and only reason this was introduced is to power the feature ci plugin status page, and this was introduced because we were asking ourself to have a distinction between default and non default plugins.

but I also see the benefits to introduce a bit more metadata for things like documentation, however I see a few things here:

* We've metadata in each plugin, aka gem, this can be used in documentation use case.
* I would rather keep this file simple, and only classification based (aka default, test, etc...) as this will give the easiest (DRY solution). 

We can always iterate, and see improvements, but for now I would rather stick to the current solution here, or also a simplification. 

What do you think?

**Comment by ph** [comment]
As the original reviewer of this PR and read the comments from others.

I usually would be in favor of @guyboertje's comment and encapsulating everything in a class and I dont think it would be overengineering stuff. **BUT**, in the other hand its only used currently used in the build process so I am OK in the quicker solution to just externalize the data.

**Comment by ph** [comment]
LGTM, lets move forward, this is a first step to dry it out.

**Comment by ph** [comment]
Can we make this move forward?

**Comment by guyboertje** [comment]
LGTM

**Comment by ph** [comment]
@purbon lets make sure we sync the content of the file, will do another round to make sure its correct.

The `LGTM` was more I am OK with the current way of making it work/

**Comment by purbon** [comment]
sure, will sync and ask you two again, I was more looking to know if there is any pending thing, a part from rebasing 👍  thanks!

**Comment by purbon** [comment]
@guyboertje @ph list updated and PR rebased, let me know what do you think and if this is good to be merged? I understand this would be master and 5.0, isn't?  or do you see it also inside other 2.x branches?

What do you think?

**Comment by ph** [comment]
@purbon I would to have this merged for master, 2.X and 5.0, I don't think it will be too hard to do it. The list of plugins is probably different between branches though.

The build is failling, but this might not be related to the current code?

**Comment by purbon** [comment]
@ph checking the build, I think the rebase got the branch screwed, fixing.

**Comment by purbon** [comment]
jenkins, please test this.

**Comment by purbon** [comment]
please jenkins, test this.

**Comment by ph** [comment]
retriggering build.

**Comment by purbon** [comment]
please jenkins, test this.

**Comment by ph** [comment]
LGTM, travis run successfully.

**Comment by ph** [comment]
@purbon I believe this is ready to merge.

**Comment by suyograo** [comment]
@purbon I can't even build the artifacts with this

```
suyog@machine:~/ws/elastic/logstash (pr/5096)$ rake artifact:tar
rake aborted!
no such file to load -- concurrent
org/jruby/RubyKernel.java:1040:in `require'
/Users/suyog/ws/elastic/logstash/logstash-core/lib/logstash/config/defaults.rb:3:in `(root)'
org/jruby/RubyKernel.java:1040:in `require'
/Users/suyog/ws/elastic/logstash/logstash-core/lib/logstash/config/cpu_core_strategy.rb:1:in `(root)'
org/jruby/RubyKernel.java:1040:in `require'
/Users/suyog/ws/elastic/logstash/logstash-core/lib/logstash/config/cpu_core_strategy.rb:3:in `(root)'
org/jruby/RubyKernel.java:1040:in `require'
/Users/suyog/ws/elastic/logstash/logstash-core/lib/logstash/environment.rb:1:in `(root)'
org/jruby/RubyKernel.java:1040:in `require'
/Users/suyog/ws/elastic/logstash/logstash-core/lib/logstash/environment.rb:3:in `(root)'
org/jruby/RubyKernel.java:1040:in `require'
/Users/suyog/ws/elastic/logstash/logstash-core/lib/logstash/json.rb:1:in `(root)'
/Users/suyog/ws/elastic/logstash/logstash-core/lib/logstash/json.rb:2:in `(root)'
org/jruby/RubyKernel.java:1059:in `load'
/Users/suyog/ws/elastic/logstash/rakelib/default_plugins.rb:1:in `(root)'
/Users/suyog/ws/elastic/logstash/rakelib/default_plugins.rb:1:in `(root)'
file:/Users/suyog/.rvm/rubies/jruby-1.7.25/lib/jruby.jar!/jruby/kernel19/kernel.rb:1:in `(root)'
file:/Users/suyog/.rvm/rubies/jruby-1.7.25/lib/jruby.jar!/jruby/kernel19/kernel.rb:13:in `require_relative'
/Users/suyog/ws/elastic/logstash/rakelib/plugin.rake:1:in `(root)'
org/jruby/RubyKernel.java:1059:in `load'
org/jruby/RubyKernel.java:1079:in `eval'
/Users/suyog/.rvm/gems/jruby-1.7.25/bin/jruby_executable_hooks:15:in `(root)'
```

**Comment by suyograo** [comment]
master build is fine. Its something with this PR.

**Comment by ph** [comment]
@suyograo I just did another test with this PR and I successfully build an artefact, but I do get a warning for redefined constants.

```
/t/logstash git:pr/5096 ❯❯❯ rake artifact:tar
/private/tmp/logstash/lib/bootstrap/environment.rb:14 warning: already initialized constant LOGSTASH_HOME
/private/tmp/logstash/lib/bootstrap/environment.rb:14 warning: already initialized constant LOGSTASH_HOME
Invoking bundler install...
```

*Notes, I did this from a clean clone*

**Comment by suyograo** [comment]
@purbon this needs a rebase with master as well.

**Comment by jordansissel** [comment]
> to add a warning blob for unsupported plugins in docs

Slight tangent: There is no concept of "unsupported plugins". "Supported" plugins is only relevant in commercial customer support relationships and is not relevant to any of the open source customers.

**Comment by ph** [comment]
Rebased in #5793
